src/charm.py

```diff
@@ -89,6 +82,14 @@ def __init__(self, *args):
             getattr(self.on, "log_data_storage_detaching"), self._on_storage_detaching
         )
 
+        self._grafana_agent = COSMachineProvider(
+            self,
+            metrics_endpoints=[
+                {"path": "/metrics", "port": 9100},
+                {"path": "/metrics", "port": 9101},
+            ],
+        )
```

Comment:
I believe one of these ports is an instance of node_exporter deployed separately. Since Grafana Agent has node exporter built in, we should not be scraping this endpoint. We should likely also remove node_exporter from the image.

Reply:
Looks like node_exporter is installed in `src/snap.py` in the `KafkaSnap` class.

Reply:
Good call @dstathis!

Currently, kafka code enables the following plugs (per [post-install instructions](https://github.com/guilhem/node_exporter-snap#post-installation)):
```python
node_exporter_plugs = [
    "hardware-observe",
    "network-observe",
    "mount-observe",
    "system-observe",
]
```
without specifying a custom list of collectors, which means that the [default ones](https://github.com/prometheus/node_exporter/blob/master/README.md#enabled-by-default) are in use. Grafana agent, however, has [a different set of collectors enabled by default](https://grafana.com/docs/agent/latest/configuration/integrations/node-exporter-config/). Afaict, the only concern is that grafana agent doesn't have `mountstats` enabled, while the snap does.
- Do we need to enable the snap plugs above with the gagent machine charm?
- Do we need to add config options to the machine charm for enabling/disabling node-exporter collectors?

Reply:
By `kafka_exporter`, do you mean https://github.com/danielqsj/kafka_exporter? 

I opened a now-parked draft PR for that exporter here - https://github.com/canonical/kafka-operator/pull/64, which uses a different port.
We're not going to implement that any time soon as it doesn't support our authentication schemes, so the only exporter that we 'have' is the JMX ones.

Reply:
The node_exporter snap we are using currently will definitely go away, as it is not maintained by Canonical, so I would not base any implementation on it. Just to clarify, we added the node_exporter snap to start to have metrics working (collecting also node_metrics) and, at that time, we were not sure about which features the grafana-agent charm/snap would ship. 

After discussing with Simon, and understanding that this functionality would have been included in the grafana agent, we decided to have it there for testing, but removing it as soon as the grafana-agent charm would be ready (which is now). 

I would just leave it up to you decide whether the mount should be enabled, although this could also be done in a separated ticket, not binded to Kafka integration I suppose

Reply:
Correction: grafana-agent's built-in node exporter and the snap have the same collectors enabled by default (see list below, rendered from the links above).
I think we're good as is at the moment (no urgency adding config options for collectors).

`arp, bcache, bonding, boottime, btrfs, conntrack, cpu, cpufreq, diskstats, dmi, edac, entropy, exec, fibrechannel, filefd, filesystem, hwmon, infiniband, ipvs, loadavg, mdadm, meminfo, netclass, netdev, netstat, nfs, nfsd, nvme, os, powersupplyclass, pressure, rapl, schedstat, selinux, sockstat, softnet, stat, tapestats, textfile, thermal, thermal_zone, time, timex, udp_queues, uname, vmstat, xfs, zfs`